src/utils/renderHelpers.ts

```diff
@@ -15,6 +15,28 @@ export interface NodeRenderOptions {
 	start?: number;
 	end?: number;
 	isNoStatement?: boolean;
+	renderedParent?: Node;
+	fieldOfRenderedParent?: string | null;
+}
+
+export function getFieldOfParent(child: Node): string | null {
+	const parent = <Node>child.parent;
+	for (const key of parent.keys) {
+		const value = (<any>parent)[key];
+		if (value === child) return key;
+		if (Array.isArray(value)) {
+			for (const nestedValue of value) {
+				if (nestedValue === child) return key;
+			}
```

Comment:
This does worry me a little - eg unnecessary lookup for long sequence expressions. This information is available to the `render` call itself, so we could just enforce it being passed for all render calls getting us back to a monomorphic render function, but I know that is a lot of work and may not even be itself desirable for performance (although maybe it would...).

If I remember correctly there are only two node types with more than one array field - template elements and tagged template elements, so that knowledge could permit an optimization here, but perhaps worth seeing if it becomes a bottleneck. I just prefer to avoid unnecessary loops that is all, and the keys loop is already unnecessary in theory due to this being parent contextual information the render function could have passed.

Reply:
Would running Rollup on uglify code that is rewritten to use long sequence expressions hit this loop?

Reply:
At the moment, expression tree-shaking is powerful but very rare, and the field will only be determined if the child is actually simplified. I do not like adding this information for every render simply because it is totally unnecessary 98% of the time. The only information currently used is the fact if it is a callee so I simplified it accordingly and removed the loop. Should we require more information at some point, we can certainly revisit that.